cpp/oneapi/dal/backend/primitives/reduction/reduction_dpc.cpp

```diff
@@ -0,0 +1,141 @@
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <type_traits>
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_dpc.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_rw_dpc.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+inline sycl::event reduce_rm_rw(sycl::queue& q,
+                                const ndview<Float, 2, ndorder::c>& input,
+                                ndview<Float, 1>& output,
+                                const BinaryOp& binary,
+                                const UnaryOp& unary,
+                                const event_vector& deps) {
+    using kernel_t = reduction_rm_rw<Float, BinaryOp, UnaryOp>;
+    const auto width = input.get_dimension(1);
+    const auto height = input.get_dimension(0);
+    const auto stride = input.get_leading_stride();
+    const auto* inp_ptr = input.get_data();
+    auto* out_ptr = output.get_mutable_data();
+    const kernel_t kernel(q);
+    return kernel(inp_ptr, out_ptr, width, height, stride, binary, unary, deps);
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+inline sycl::event reduce_rm_cw(sycl::queue& q,
+                                const ndview<Float, 2, ndorder::c>& input,
+                                ndview<Float, 1>& output,
+                                const BinaryOp& binary,
+                                const UnaryOp& unary,
+                                const event_vector& deps) {
+    using kernel_t = reduction_rm_cw<Float, BinaryOp, UnaryOp>;
+    const auto width = input.get_dimension(1);
+    const auto height = input.get_dimension(0);
+    const auto stride = input.get_leading_stride();
+    const auto* inp_ptr = input.get_data();
+    auto* out_ptr = output.get_mutable_data();
+    const kernel_t kernel(q);
+    return kernel(inp_ptr, out_ptr, width, height, stride, binary, unary, deps);
+}
+
+template <typename Float, ndorder order, typename BinaryOp, typename UnaryOp>
+sycl::event reduce_rows(sycl::queue& q,
+                        const ndview<Float, 2, order>& input,
+                        ndview<Float, 1>& output,
+                        const BinaryOp& binary,
+                        const UnaryOp& unary,
+                        const event_vector& deps) {
+    ONEDAL_ASSERT(input.get_dimension(0) <= output.get_dimension(0));
```

Comment:
Didn't get this assertion. The count of rows in the input must be less than in output?

Reply:
It mean that output array have enough room to store all reduced cols